File: AngularNetCoreProj/ClientApp/src/app/product/basket.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, OnInit } from '@angular/core';
import { ToastrService } from 'ngx-toastr';
import { BehaviorSubject, Observable } from 'rxjs';
import { v4 as uuidv4 } from 'uuid';
import { environment } from '../../environments/environment';
import { IBasket } from '../entity/IBasket';
import { IBasketTotal } from "../entity/IBasketTotal";
import { IBasketItem } from '../entity/IBasketItem';
import { IProduct } from '../entity/IProduct';
import { map } from 'rxjs/operators';

@Injectable({ providedIn: "root" })
export class BasketService implements OnInit
{
 
  public basket: IBasket;
  private basketid: string;
  private basketSource = new BehaviorSubject<IBasket>(null);
  public basketTotalSource = new BehaviorSubject<IBasketTotal>(null);
  basket$ = this.basketSource.asObservable();
  basketTotal$ = this.basketTotalSource.asObservable();

  constructor(private http: HttpClient) {
   
    this.basketid = this.getBasketId();
    this.basket = new IBasket();
    this.basket.basketid = this.basketid;
    this.getBasket();
  
  }

  ngOnInit(): void {
    
  }

  getBasketId(): string {
    let id = localStorage.getItem("basketid");
    if (id == null) {
      id = uuidv4();
      localStorage.setItem("basketid", id);
    }
    return id;

  }

  getBasket() {
    this.http.get<(IBasket)>(environment.apiurlBasket + "/" + this.basketid)
      .pipe(
        map((response: IBasket) => {
          if (response != null) {
            this.basketSource.next(response);
            this.basket = response;
            this.calculateTotalQuantity(response);
          }}
        )
    ).subscribe()
   }

  calculateTotalQuantity(basket: IBasket) {
    let totalquantity: number = basket.items.reduce((a, b) => a + b.quantity, 0);
    let baskettotal: IBasketTotal = new IBasketTotal();
    baskettotal.total = totalquantity ;
    this.basketTotalSource.next(baskettotal);

  }
  setBasket() {
    return this.http.post("http://localhost:5002/api/basket/", this.basket).pipe(
      map(
      sucess => {
        this.basketSource.next(this.basket);
        this.calculateTotalQuantity(this.basket);
      }
    ));
  }

  deleteBasket() {
    let url = "http://localhost:5002/api/basket/" + this.getBasketId();
    return this.http.delete(url).pipe(
      map(
      sucess => {
        this.basket = new IBasket();
        this.basketSource.next(this.basket);
        this.calculateTotalQuantity(this.basket);
      }
    ));
  }

  updateBasketItem(basketItem: IBasketItem)
  {
    var temp = this.basket.items.filter((e) => e.id == basketItem.id);
    if (temp.length > 0) {
      temp[0]=basketItem;
    }
    else {
      basketItem.quantity = 1;
      this.basket.items.push(basketItem)
    };
    return this.setBasket();
  }

  removeItemFromBasket(item: IBasketItem) {
    //write code to remove the item from the basket.items collection
    this.basket.items =     this.basket.items.filter((e) => e.id != item.id); 
    return this.setBasket();  
  }   
 

  getBasketItem(basketitemid: number) {
    var temp = this.basket.items.filter((e) => e.id == basketitemid);
    if (temp.length > 0) {
      return temp[0];
    }
    else
    {return null;}
  }

  

  ProductToBasketItem(product: IProduct): IBasketItem
  {
    return {
      id: product.id,
      name : product.name,
      price : product.price,
      productCategory : product.productCategory.description,
      productCategoryId: product.productCategoryId,
      quantity : 0
    };
  }
    
}

File: AspnetMVC/Services/BasketService.cs
using AspnetMVC.Models;
using Newtonsoft.Json;

namespace AspnetMVC.Services
{
    public class BasketService : IDisposable, IBasketService
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;
        private readonly string _basketId;

        public BasketService(HttpClient http)
        {
            _httpClient = http;
            _baseUrl = "http://localhost:5002/api/basket/";
            _basketId = GetBasketId();
        }

        public async Task<Basket> GetBasket()
        {
            var response = await _httpClient.GetAsync(_baseUrl + _basketId);
            response.EnsureSuccessStatusCode();
            var json = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<Basket>(json);
        }

        public async Task SetBasket(Basket basket)
        {
            var json = JsonConvert.SerializeObject(basket);
            var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
            var response = await _httpClient.PostAsync(_baseUrl, content);
            response.EnsureSuccessStatusCode();
        }

        public async Task DeleteBasket()
        {
            var response = await _httpClient.DeleteAsync(_baseUrl + _basketId);
            response.EnsureSuccessStatusCode();
        }

        public async Task<Basket> UpdateBasketItem(BasketItem basketItem)
        {
            var basket = await GetBasket();
            if (basket == null)
            { 
                basket = new Basket();
                basket.basketid = GetBasketId();
               
            }

            var existingItem = basket.items.Find(item => item.id == basketItem.id);
            if (existingItem != null)
            {
                existingItem.quantity = existingItem.quantity + 1;
            }
            else
            {
                basketItem.quantity = 1;
                basket.items.Add(basketItem);
            }
            await SetBasket(basket);
            return basket;
        }

        public async Task RemoveItemFromBasket(BasketItem item)
        {
            var basket = await GetBasket();
            basket.items.RemoveAll(i => i.id == item.id);
            await SetBasket(basket);
        }

        public BasketItem ProductToBasketItem(Product product)
        {
            return new BasketItem
            {
                id = product.Id,
                name = product.Name,
                price = product.Price,
                productCategory = product.ProductCategory.Description,
                productCategoryId = product.ProductCategoryId,
                quantity = 0
            };
        }

        private string GetBasketId()
        {
            //write code to fetch cookie value


            var id = "matt_basket";
            //if (id == null)
            //{
            //    id = Guid.NewGuid().ToString();
            //    localStorage.setItem("basketid", id);
            //}
            return id;
        }

        public void Dispose()
        {
            _httpClient.Dispose();
        }

        public int GetBasketQuantity(Basket basket)
        {
            return basket.items.Sum<BasketItem>(e=>e.quantity);
        }

        

      
    }

}

